cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -194,10 +195,21 @@ protected void processDag(PhaseSpec phaseSpec, String sourcePluginType, JavaSpar
     //Emitted records and sinkRunnables will be populated as each stage is processed
     SinkRunnableProvider sinkRunnableProvider = new BatchSinkRunnableProvider();
     for (String stageName : groupedDag.getTopologicalOrder()) {
-      processStage(phaseSpec, sourcePluginType, sec, stagePartitions, pluginContext, collectors,
+      try {
+        processStage(phaseSpec, sourcePluginType, sec, stagePartitions, pluginContext, collectors,
           pipelinePhase, functionCacheFactory, macroEvaluator, emittedRecords, groupedDag, groups,
           branchers, shufflers, sinkRunnables, stageName, System.currentTimeMillis(), null,
           sinkRunnableProvider);
+      } catch (Exception e) {
+        List<Throwable> causalChain = Throwables.getCausalChain(e);
+        for (Throwable t : causalChain) {
+          if (t instanceof WrappedStageException) {
```

Comment:
if everything is wrapped, when would WrappedStageException not be in the causal chain?

Reply:
There are cases for example when we do `joins`, it will do `handleAutoJoin->SparkCollection#join` where it does not execute the code paths which is wrapped.

Reply:
In those cases, pipeline can fail during processing of a particular join stage but it does not throw `WrappedStageException`